### ISSUE TITLE
Graphormer module

### ISSUE THREAD

**Issue by paridhimaheshwari2708** [microsoft/Graphormer#133]
How can I use the graphormer model with custom dataloader and training scripts (not the fairseq commands)?
I want to use Graphormer as a torch.nn.Module (like any other GNN layer).

**Comment by mavisguan** [comment]
Hi! Given our current implementation, it might take some extra effort to satisfy your need. You can try to wrap up code directly related to Graphormer model (which is scattered among 4~5 files, like `graphormer/models/graphormer.py`, `graphormer/tasks/graph_prediction.py`, etc.) into a single file. For this, you can also refer to fairseq's `build_model` funtion implementation. Then you can import it as a normal python module.

**Comment by paridhimaheshwari2708** [comment]
@mavisguan Hi, thank you for your suggestion. Could you also provide more information about the dataloader? I have a custom dataset of DGL graphs and a task-specific sampling that happens in the dataloader. How can I wrap it into the format needed by Graphormer? Specifically, what does `batched_data` ([input to graphormer](https://github.com/microsoft/Graphormer/blob/main/graphormer/models/graphormer.py#L151)) look like?

**Comment by mavisguan** [comment]
@paridhimaheshwari2708 You can use an example DGL dataset (like qm7b) and put a breakpoint before the line of code you've mentioned, and see how `batched_data` looks like:
![image](https://user-images.githubusercontent.com/58137527/185039546-ca2839c8-dab0-4826-96db-cc0a6cb974de.png)

**Comment by paridhimaheshwari2708** [comment]
@mavisguan Thank you, this is really helpful! I noticed [here](https://github.com/microsoft/Graphormer/blob/3de5d51cd5880a710a35a66a2660654057925a2e/graphormer/data/dgl_datasets/dgl_dataset.py) that DGL graphs are actually converted into PyG graphs. Is that right? If so, is `batched_data` obtained from `torch_geometric.loader.DataLoader`.

**Comment by mavisguan** [comment]
Yes, you're right. I think `batched_data` is obtained from `torch_geometric.loader.DataLoader`, which contains `batch_size` PYG graphs, and it's also a dictionary.
![image](https://user-images.githubusercontent.com/58137527/185283882-8530b155-9e0d-4f37-8b4d-6f3e9577df3a.png)

**Comment by laowu-code** [comment]
> Yes, you're right. I think `batched_data` is obtained from `torch_geometric.loader.DataLoader`, which contains `batch_size` PYG graphs, and it's also a dictionary. ![image](https://user-images.githubusercontent.com/58137527/185283882-8530b155-9e0d-4f37-8b4d-6f3e9577df3a.png)
Sorry to bother you,I am wondering how can I convert my dataset(not the dgl,obg or pyg,but the graph adj)into the type that fits batched_data,thanks a lot.

**Comment by mavisguan** [comment]
@laowu-code You can try to convert your dataset into PYG's customized dataset, following their official document https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html

**Comment by laowu-code** [comment]
@mavisguan  Thanks for your reply,I will try it,it means a lot to me.

**Comment by engmubarak48** [comment]
Hi, 


The PyTorch geometric loader holds node features as (N * feat_dim), where N could be the number of nodes of all graphs in your batch. And in your GraphNodeFeature forward [here](https://github.com/microsoft/Graphormer/blob/77f436db46fb9013121289db670d1a763f264153/graphormer/modules/graphormer_layers.py#L53), your node features has ([n_graph, n_node, feat_dim]). Could you please refer to me the line where you do this sort of batching? Do you pad it by the maximum number of nodes in the graph? Or am I missing something here?

**Comment by BrandenKeck** [comment]
> @paridhimaheshwari2708 You can use an example DGL dataset (like qm7b) and put a breakpoint before the line of code you've mentioned, and see how `batched_data` looks like: ![image](https://user-images.githubusercontent.com/58137527/185039546-ca2839c8-dab0-4826-96db-cc0a6cb974de.png) I use this debugging toolkit to set breakpoints in big python projects: https://github.com/volltin/vpack. I think it's a very helpful tool for digging into Graphormer's code. We're sorry that our tutorials on customizing datasets is incomplete and a bit ambiguous, and we're working on updating our tutorials, please stay tuned.

Hi mavisguan,

Is it possible to construct this dictionary manually without a dataloader to pass to the model for testing purposes (just to ensure that the modules are setup correctly)?  For example:

`batched_data = {
    "idx": ,
    "x": ,
    "attn_bias": ,
    "attn_edge_type": ,
    "spatial_pos": ,
    "degree": ,
    "edge_input": ,
    "y": 
}`

If so, what format is each component?  And, does this need to be wrapped as TensorDict, or is it a dictionary of tensors?  I just want to load one molecule example to see if I'm doing this correctly - I modified the code to combine "in degree" and "out degree" and remove multi-hop.  However, I am becoming confused by this line:

`node_feature = self.atom_encoder(x).sum(dim=-2)`

I would have expected the 'x' key to be an adjacency matrix, but it appears that this information should include node types for the encoder?  But, the same dictionary key is passed to multi-head attention, so I think I'm confusing myself.

**Comment by BrandenKeck** [comment]
For anyone that comes across this that is also struggling with the data structure - I think I understand now after reading [wrapper.py](https://github.com/microsoft/Graphormer/blob/main/graphormer/data/wrapper.py).  "batched_data" appears to be a python dictionary with tensors of the following shapes:

- **idx**: `(batch size)` dataset indices
- **x**: `(batch size, num atoms, num atom features)` where the features are atomic number, chirality, etc. (there is an example from torch geometric with these features [here](https://pytorch-geometric.readthedocs.io/en/2.5.2/_modules/torch_geometric/utils/smiles.html) )
- **edge_input**: `(batch size, num atoms, num atoms, max dist. between atoms, num bond features)` contains the path information from shortest paths.  I'm still trying to understand it, but it can be generated from the code in the repo without a full understanding
- **attn_bias**: `(batch size, num atoms + 1, num atoms + 1)` is initialized to zeros with an extra atom dimension presumably for the virtual node
- **attn_edge_type**: `(batch size, num atoms, num atoms, num bond features)` is simply a list of edge types in somewhat of an adjacency matrix format
- **spacial_pos**: `(batch size, num atoms, num atoms)` is the shortest path length between each atom and every other atom
- **degree**: `(batch size, num atoms)` is the degree of each node in the graph.  I modified wrapper.py and the graphformer code to include only "degree" as I am testing this on undirected graphs.  The wrapper.py code sets in_degree = out_degree by default if both are included.
